Documentation/teaching/lectures/syscalls.rst

```diff
@@ -0,0 +1,596 @@
+system call numbers to kernel functions:
+
+.. slide:: System Call Table
+   :inline-contents: True
+   :level: 2
+
+   .. code-block:: c
+
+      #define __SYSCALL_I386(nr, sym, qual) [nr] = sym,
+
+      const sys_call_ptr_t ia32_sys_call_table[] = {
+        [0 ... __NR_syscall_compat_max] = &sys_ni_syscall,
+        #include <asm/syscalls_32.h>
+      };
+
+   .. code-block:: c
+
+      __SYSCALL_I386(0, sys_restart_syscall, )
+      __SYSCALL_I386(1, sys_exit, )
+      #ifdef CONFIG_X86_32
+      __SYSCALL_I386(2, sys_fork, )
+      #else
+      __SYSCALL_I386(2, sys_fork, )
+      #endif
+      __SYSCALL_I386(3, sys_read, )
+      __SYSCALL_I386(4, sys_write, )
+
+
+
+System call parameters handling
+-------------------------------
+
+Handling system calls parameters is tricky. Since these values are
+setup by userspace, the kernel can not assume correctness and must
+always verify them throughly.
+
+Pointers have a few important special case that must be checked:
+
+.. slide:: System Calls Pointer Parameters
+   :inline-contents: True
+   :level: 2
+
+   * Never allow pointers to kernel-space
+
+   * Check for invalid pointers
+
+
+Since system calls are executed in kernel mode, they have access to
+kernel space and if pointers are not properly checked user
+applications might get read or write access to kernel space.
+
+For example, lets consider the case where such a check is not made for
+the read or write system calls. If the user passes a kernel-space
+pointer to a write sysytem call then it can get access to kernel data
+by later reading the file. If it passes an kernel-space pointer to a
+read system call then it can corrupt kernel memory.
+
+
+.. slide:: Pointers to Kernel Space
+   :level: 2
+
+   * User access to kernel data if allowed in a write system call
+
+   * User corrupting kernel data if allowed in a read system call
+
+
+Likewise, if a pointer passed by the application is invalid
+(e.g. unmaped, read-only for cases where its need to be written), it
+could "crash" the kernel. There two approaches that could be used:
+
+.. slide:: Invalid pointers
+   :inline-contents: True
+   :level: 2
+
+   * Check the pointer against the user address space
+
+   * Rely on the MMU to avoid checking the pointer for the case when
+     it is correct and handle the fault for the invalid case
+
+Although it sounds tempting, the second approach is not that easy to
+implement. The page fault handler uses the fault and faulting
+addresses as well as information from the user address space to
+determine the cause:
+
+.. slide:: Page fault handling
+   :inline-contents: True
+   :level: 2
+
+      * Copy on write, demand paging, swapping: both the fault and
+	faulting addresses are in userspace; the fault address is
+	valid (checked against the user address space)
+
+      * Invalid syscall pointer: the faulting address is in kernel
+	space; the fault address is in userspace and it is invalid
+
```

Comment:
I also don't get "invalid syscall pointer" bullet.  :)

So do you mean invalid user space pointer that is accessed from kernel.